providers/globus.go

```diff
@@ -0,0 +1,171 @@
+package providers
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+)
+
+type GlobusProvider struct {
+	*ProviderData
+}
+
+type Token struct {
+	AccessToken string  `json:"access_token"`
+	ExpiresIn   int     `json:"expires_in"`
+	Scope       string  `json:"scope"`
+	Tokens      []Token `json:"other_tokens"`
+}
+
+//https://auth.globus.org/.well-known/openid-configuration
+func NewGlobusProvider(p *ProviderData) *GlobusProvider {
+	p.ProviderName = "Globus"
+	if p.LoginURL == nil || p.LoginURL.String() == "" {
+		p.LoginURL = &url.URL{
+			Scheme: "https",
+			Host:   "auth.globus.org",
+			Path:   "/v2/oauth2/authorize",
+		}
+	}
+	if p.RedeemURL == nil || p.RedeemURL.String() == "" {
+		p.RedeemURL = &url.URL{
+			Scheme: "https",
+			Host:   "auth.globus.org",
+			Path:   "/v2/oauth2/token",
+		}
+	}
+	if p.ValidateURL == nil || p.ValidateURL.String() == "" {
+		p.ValidateURL = &url.URL{
+			Scheme: "https",
+			Host:   "auth.globus.org",
+			Path:   "/v2/oauth2/token/introspect",
+		}
+	}
+	if p.ProfileURL == nil || p.ProfileURL.String() == "" {
+		p.ProfileURL = &url.URL{
+			Scheme: "https",
+			Host:   "auth.globus.org",
+			Path:   "/v2/oauth2/userinfo",
+		}
+	}
+	if p.Scope == "" {
+
+		// TODO:  scopes should be configurable
+		p.Scope = "openid email profile urn:globus:auth:scope:auth.globus.org:view_identities urn:globus:auth:scope:transfer.api.globus.org:all urn:globus:auth:scope:data.materialsdatafacility.org:all urn:globus:auth:scope:transfer.api.globus.org:all urn:globus:auth:scope:search.api.globus.org:search"
+
+	}
+	return &GlobusProvider{ProviderData: p}
+}
+
+func (p *GlobusProvider) GetEmailAddress(c context.Context, s *sessions.SessionState) (string, error) {
+	var userinfo struct {
+		Email             string `json:"email"`
+		Name              string `json:"name"`
+		PreferredUsername string `json:"preferred_username"`
+		Sub               string `json:"sub"`
+	}
+
+	req, _ := http.NewRequest("GET", p.ProfileURL.String(), nil)
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", s.AccessToken))
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", err
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	resp.Body.Close()
+	if err != nil {
+		return "", err
+	}
+
+	if resp.StatusCode != 200 {
+		return "", fmt.Errorf("got %d from %q %s",
+			resp.StatusCode, p.ProfileURL.String(), body)
+	} else {
+		log.Printf("got %d from %q %s", resp.StatusCode, p.ProfileURL.String(), body)
+	}
+
+	if err := json.Unmarshal(body, &userinfo); err != nil {
+		return "", fmt.Errorf("%s unmarshaling %s", err, body)
+	}
+
+	return userinfo.Email, nil
+}
+
+func (p *GlobusProvider) Redeem(c context.Context, redirectURL string, code string) (s *sessions.SessionState, err error) {
+	if code == "" {
+		err = errors.New("missing code")
+		return
+	}
+
+	params := url.Values{}
```

Comment:
Take a look at other implementations of the OAuth2 redeem flow -- we've refactored alot of them and have custom libraries in the codebase to manage alot of this now (e.g. request builder).

Does this have any non-standard aspects, otherwise you should use the default `Redeem` functionality in `provider_defaults`.

You can look at `keycloak.go` as an example of a provider that only customized the `EnrichSession` method and could use the defaults for everything else.